ray_julia_jll/gen/artifacts.jl

```diff
@@ -0,0 +1,133 @@
+using Base: SHA1, BinaryPlatforms
+using CodecZlib: GzipCompressorStream, GzipDecompressorStream
+using CURL_jll
+using jq_jll
+using LibGit2: LibGit2
+using Pkg.Artifacts: bind_artifact!
+using Pkg.Types: read_project
+using SHA: sha256
+using Tar
+using wget_jll
+
+# TODO: tidy up
+const TRIPLET_REGEX = r"""
+    ^ray_julia.v(?<jll_version>[0-9]\.[0-9]\.[0-9])\.
+    (?<triplet>[a-z, 0-9, \-, \_]+)
+    -julia_version\+(?<julia_version>[0-9]\.[0-9]\.[0-9])
+    .tar.gz$
+    """x
+
+const GH_RELEASE_ASSET_PATH_REGEX = r"""
+^/(?<owner>[^/]+)/(?<repo_name>[^/]+)/
+releases/download/
+(?<tag>[^/]+)/(?<file_name>[^/]+)$
+"""x
+# Compute the Artifact.toml `git-tree-sha1`.
+function tree_hash_sha1(tarball_path)
+    return open(GzipDecompressorStream, tarball_path, "r") do tar
+        SHA1(Tar.tree_hash(tar))
+    end
+end
+
+# Compute the Artifact.toml `sha256` from the compressed archive.
+function sha256sum(tarball_path)
+    return open(tarball_path, "r") do tar
+        bytes2hex(sha256(tar))
+    end
+end
+
+function remote_url(repo_root::AbstractString, name::AbstractString="origin")
+    return LibGit2.with(LibGit2.GitRepo(repo_root)) do repo
+        LibGit2.with(LibGit2.lookup_remote(repo, name)) do remote
+            LibGit2.url(remote)
+        end
+    end
+end
+
+function get_release_asset_urls(jll_version)
+    # TODO: parse URL from pkg_url
+    assets_url = "https://api.github.com/repos/beacon-biosignals/Ray.jl/releases/tags/v$(jll_version)"
+    io = IOBuffer()
+    run(pipeline(`$(curl()) $assets_url`, `$(jq()) -r '.assets[].browser_download_url'`, io))
+    assets = split(String(take!(io)), "\n"; keepempty=false)
+    return assets
+end
+
+function download_asset(asset_url, dir)
+    run(`$(wget()) $asset_url -P $dir`)
+    filename = replace(basename(asset_url), "%2B" => "+") # TODO: better way to parse URL in unicode?
```

Comment:
any suggestions welcome - tried doing it via `curl` and `jq` but I feel like there should be a nice Julian way to do this?

Reply:
Use the `URIs` package to do the escaping:
```
julia> using URIs

julia> URIs.escapeuri("0.1.2+3")
"0.1.2%2B3"

julia> URIs.unescapeuri("0.1.2%2B3")
"0.1.2+3"
```
That said, you should just use the `Downloads` package or the buildin `download` function to handle this and I doubt you'll even need to do escaping